File: algoritmos_e_estruturas/Projetos_de_algoritmos/Programação_Dinamica/fibonacci/fibonacci_button_up.cpp
#include <iostream>
#define TAM 100

using namespace std;


int fibonacci_button_up(int n){
    /*Fibonacci iterativo, comecando de um subproblema menor até o maior*/

   int vetor[TAM];
   for (int i= 0; i<TAM;i++){
    vetor[i] = -1;
   }

   vetor[0] = 0;
   vetor[1] = 1;
   vetor[2] = 1;

   for (int i=3;i <=n ;i++){
        vetor[i] = vetor[i-1] + vetor[i-2];
   }
   return vetor[n];
}

int main(){
    /*
    Toda vez que um problema é resolvido é salvo no array.
    return 1, se n == 1 ou n == 2
    return Fib(n-1) + Fib(n-2), se n > 2
    */
   cout << "Fibonacci de " << 6 << ": " << fibonacci_button_up(6) << endl;
   cout << "Fibonacci de " << 7 << ": " << fibonacci_button_up(7) << endl;   
   cout << "Fibonacci de " << 8 << ": " << fibonacci_button_up(8) << endl;   

  
    return 0;
}

File: algoritmos_e_estruturas/Projetos_de_algoritmos/Programação_Dinamica/fibonacci/fibonacci_button_up.py
"Fibonacci top down - iterativo -> do menor para o maior"

def fibonacci_botton_up(n:int):
    # 1, 1, 2, 3, 5, 8, 13, 21, 34, 55, 89

    if n <= 2:
        return 1
    
    fib: dict = {}
    fib[1] = 1
    fib[2] = 1
    for i in range(3, n+1):
        fib[i] = fib[i-1] + fib[i-2]
    
    return fib[n]

if __name__ == "__main__":
    # 1  2  3  4  5  6  7   8   9   10  11   
    # 1, 1, 2, 3, 5, 8, 13, 21, 34, 55, 89
    print(f'fibonacci: 4: ', fibonacci_botton_up(4))
    print(f'fibonacci: 7: ', fibonacci_botton_up(7))
    print(f'fibonacci: 8: ', fibonacci_botton_up(8))
    print(f'fibonacci: 9: ', fibonacci_botton_up(9))
    print(f'fibonacci: 10: ',fibonacci_botton_up(10))
